File: musicstudio/dsp/src/musicstudio_dsp/waveform.py
"""Waveform reduction for Requirement 12.7.

A drawing needs the extreme of each bucket, not the samples: a five-minute asset is
14.4 million frames and a client draws a few hundred pixels. What the product layer
decided — two extremes per bucket, normalised to [-1, 1] — is recorded in
``domain/playback/waveform.ts``; this module is the reduction itself.

The bucket boundaries are computed here the same way that module computes them, and
``test/test_waveform.py`` pins the agreement: frames rarely divide evenly into buckets and
the remainder is spread over the first buckets rather than piled onto the last, which
would make the final bucket of a long asset visibly taller for a reason that is arithmetic
rather than audio.

Deterministic, like every other reduction in this package: no RNG, no thread count, and a
fixed traversal order, so two calls on one asset give the same drawing.
"""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np

from .audio_buffer import AudioBuffer

__all__ = [
    "WAVEFORM_BUCKETS_MAX",
    "WAVEFORM_BUCKETS_MIN",
    "WaveformBucket",
    "bucket_boundaries",
    "reduce_to_waveform",
]

#: Mirrors ``domain/playback/waveform.ts``. `test_waveform.py` fails if the two drift.
WAVEFORM_BUCKETS_MIN = 16
WAVEFORM_BUCKETS_MAX = 4_000


@dataclass(frozen=True)
class WaveformBucket:
    """One bucket's extremes, in [-1, 1]."""

    min: float
    max: float


def bucket_boundaries(frame_count: int, buckets: int) -> list[int]:
    """Where each bucket begins, plus a final entry at ``frame_count``.

    The same spreading rule as ``bucketBoundaries`` in the TypeScript module: the first
    ``frame_count % buckets`` buckets take one extra frame.
    """
    if frame_count <= 0 or buckets <= 0:
        return []

    base, remainder = divmod(frame_count, buckets)
    boundaries: list[int] = []
    cursor = 0
    for index in range(buckets):
        boundaries.append(cursor)
        cursor += base + (1 if index < remainder else 0)
    boundaries.append(frame_count)
    return boundaries


def reduce_to_waveform(audio: AudioBuffer, buckets: int) -> list[WaveformBucket]:
    """Reduce every channel to ``buckets`` pairs of extremes.

    Channels are folded together — the extreme of a bucket is the extreme across all of
    them — because Requirement 12.7 asks for *the asset's* waveform and a two-channel
    drawing would need the client to decide what to do with the second. The fold is a
    maximum rather than a mean, so a transient present in one channel stays visible.

    ``buckets`` is clamped to the frame count: a bucket per frame is the sample dump this
    reduction exists to avoid, and asking for more buckets than frames cannot produce more
    information than the audio holds.
    """
    if not audio.is_well_formed:
        raise ValueError("input audio is not well formed")

    frames = audio.frame_count
    resolved = max(1, min(int(buckets), frames))
    boundaries = bucket_boundaries(frames, resolved)

    # One stacked view rather than a per-channel loop with a merge afterwards: the fold is
    # then a single reduction over an axis, which is both faster and impossible to get
    # inconsistent between the two extremes.
    stacked = np.stack(audio.channels, axis=0)

    result: list[WaveformBucket] = []
    for index in range(resolved):
        start = boundaries[index]
        end = boundaries[index + 1]
        if end <= start:
            result.append(WaveformBucket(min=0.0, max=0.0))
            continue

        window = stacked[:, start:end]
        result.append(
            WaveformBucket(min=float(np.min(window)), max=float(np.max(window)))
        )

    return result

File: musicstudio/domain/playback/waveform.ts
/**
 * Waveform data (Requirement 12.7), and where its numbers come from.
 *
 * Requirement 12.7 says "파형 데이터를 반환한다" and fixes nothing else, so the shape is a
 * product decision, recorded here:
 *
 * - **Peaks, not samples.** A five-minute asset is 14.4 million frames; a client draws a few
 *   hundred pixels. What a drawing needs is the extreme of each bucket, and sending frames
 *   would move megabytes to render kilobytes.
 * - **Both extremes per bucket.** A waveform drawn from `max(abs)` alone is symmetric and
 *   loses the asymmetry real audio has. Two numbers per bucket cost nothing next to the
 *   saving above.
 * - **Normalised to [-1, 1].** The same range the samples are in (design §5.1), so a client
 *   scales by its own height and nothing has to know the bit depth.
 *
 * The reduction itself runs in the DSP worker (`dsp/src/musicstudio_dsp/waveform.py`); this
 * module is the bucket arithmetic both sides agree on, so a cached waveform computed at one
 * bucket count can be checked against a request for another.
 */

export const WAVEFORM_BUCKETS_MIN = 16;
export const WAVEFORM_BUCKETS_MAX = 4_000;
export const WAVEFORM_BUCKETS_DEFAULT = 800;

export interface WaveformBucket {
  /** Most negative sample in the bucket, in [-1, 0]. */
  readonly min: number;
  /** Most positive sample in the bucket, in [0, 1]. */
  readonly max: number;
}

export interface Waveform {
  readonly assetId: string;
  readonly buckets: readonly WaveformBucket[];
  /** What the buckets span, so a client can map a bucket to a time. */
  readonly durationMs: number;
  readonly channels: number;
}

export function isWaveformBucketCount(value: unknown): boolean {
  return (
    typeof value === 'number' &&
    Number.isInteger(value) &&
    value >= WAVEFORM_BUCKETS_MIN &&
    value <= WAVEFORM_BUCKETS_MAX
  );
}

/**
 * How many frames fall in each bucket, and where each begins.
 *
 * Frames rarely divide evenly into buckets, and the remainder has to go somewhere. It is
 * spread — the first `frames % buckets` buckets take one extra — rather than piled onto the
 * last one, which would make the final bucket of a long asset visibly taller than its
 * neighbours for a reason that is arithmetic rather than audio.
 */
export function bucketBoundaries(frameCount: number, buckets: number): readonly number[] {
  if (frameCount <= 0 || buckets <= 0) return [];

  const base = Math.floor(frameCount / buckets);
  const remainder = frameCount % buckets;
  const boundaries: number[] = [];

  let cursor = 0;
  for (let index = 0; index < buckets; index += 1) {
    boundaries.push(cursor);
    cursor += base + (index < remainder ? 1 : 0);
  }
  boundaries.push(frameCount);
  return boundaries;
}

/**
 * The time a bucket starts at, for Requirement 12.5's synchronisation and for a client
 * mapping a click on the waveform back to a position.
 */
export function bucketStartMs(index: number, buckets: number, durationMs: number): number {
  if (buckets <= 0) return 0;
  return (Math.max(0, Math.min(index, buckets)) * durationMs) / buckets;
}

/** Fewer buckets than frames, always: a bucket per frame is the sample dump 12.7 is not. */
export function resolveBucketCount(requested: number | undefined, frameCount: number): number {
  const wanted = isWaveformBucketCount(requested) ? (requested as number) : WAVEFORM_BUCKETS_DEFAULT;
  return Math.max(1, Math.min(wanted, frameCount));
}
